almost_tcp/message_hdl.py

```diff
@@ -0,0 +1,249 @@
+    # The "syn" message is considered to occupy the first octet of the stream.
+    seq: 16
+    # Acknowledgement number: the next number the sender of this ACK expects.
+    ack: 16
+
+
+class PacketSignature(Signature):
+    """
+    Signature of a packet producer.
+
+    Attributes
+    ----------
+    header:         Header, out
+                    The header curently read in to the buffer.
+    stream_valid:   Out(1)
+                    High if the "stream" value from the header is valid.
+    header_valid:   Out(1)
+                    High if the whole header is valid.
+
+    data:           Stream(8), out
+                    Output stream for the body of the packet.
+                    The valid bits automatically deassert after the body
+                    has been fully read from the data stream.
+    """
+
+    def __init__(self):
+        super().__init__({
+            "input": In(stream.Signature(8)),
+            "header": Out(HeaderLayout),
+            "stream_valid": Out(1),
+            "header_valid": Out(1),
+            "data": Out(stream.Signature(8)),
+        })
+
+
+class ReadPacketStop(Component):
+    """
+    Stop on a packet-reading bus.
+
+    Each stream should have a stop on the same bus,
+    chained by their "inbus" and "outbus" connections.
+    When a stop spots a packet for its stream,
+    it tees off the input into its own PacketSignature output.
+
+    All stops get all bytes, even those not destined for this stop.
+
+    Parameters
+    ----------
+    id: int
+        Stream ID for this stop.
+
+    Attributes
+    -----------
+    inbus: Out(Stream(8))
+        Input from the bus.
+    outbus: In(Stream(8))
+        Output to the next stop on the bus.
+    packet:   Out(PacketSignature)
+        The currently-buffered packet.
+    """
+
+    inbus: In(stream.Signature(8))
+    outbus: Out(stream.Signature(8))
+    packet: Out(PacketSignature())
+
+    def __init__(self, id: int):
+        super().__init__()
+        self._stream_id = id
+
+    def elaborate(self, platform):
+        m = Module()
+
+        # Our header may be locally valid
+        # but not a match for the stream.
+        local_header_valid = Signal(1)
+        header_valid_and_matched = self.packet.header_valid
+        stream_valid = self.packet.stream_valid
+        data = self.packet.data
+        stream_match = Signal(1)
+
+        # We add a 1-slot FIFO to the bus output
+        # to avoid long combinatorial paths.
+        m.submodules.outbus = outbus = SyncFIFOBuffered(width=8, depth=1)
+        m.d.comb += [
+            self.outbus.payload.eq(outbus.r_data),
+            self.outbus.valid.eq(outbus.r_rdy),
+            outbus.r_en.eq(self.outbus.ready),
+        ]
+        # We don't do the same on the "local data" side;
+        # we assume the reader of the body will have its own
+        # (connection-level) buffer.
+
+        # Our state machine: what byte do we read in to next?
+        byte = Signal(4)
```

Comment:
Weird to call this byte, but have it be a Signal(4)? Not really a problem, though.

Reply:
Renamed as `byte_counter`, since that's what it's doing.